components/login/login_controller/login_controller.tsx

```diff
@@ -31,45 +28,60 @@ import SuccessIcon from 'components/widgets/icons/fa_success_icon';
 import WarningIcon from 'components/widgets/icons/fa_warning_icon';
 import LocalizedInput from 'components/localized_input/localized_input';
 import Markdown from 'components/markdown';
-
 import LoginMfa from '../login_mfa.jsx';
 
-class LoginController extends React.PureComponent {
-    static propTypes = {
-        intl: intlShape.isRequired,
-
-        location: PropTypes.object.isRequired,
-        isLicensed: PropTypes.bool.isRequired,
-        currentUser: PropTypes.object,
-        customBrandText: PropTypes.string,
-        customDescriptionText: PropTypes.string,
-        enableCustomBrand: PropTypes.bool.isRequired,
-        enableLdap: PropTypes.bool.isRequired,
-        enableOpenServer: PropTypes.bool.isRequired,
-        enableSaml: PropTypes.bool.isRequired,
-        enableSignInWithEmail: PropTypes.bool.isRequired,
-        enableSignInWithUsername: PropTypes.bool.isRequired,
-        enableSignUpWithEmail: PropTypes.bool.isRequired,
-        enableSignUpWithGitLab: PropTypes.bool.isRequired,
-        enableSignUpWithGoogle: PropTypes.bool.isRequired,
-        enableSignUpWithOffice365: PropTypes.bool.isRequired,
-        experimentalPrimaryTeam: PropTypes.string,
-        ldapLoginFieldName: PropTypes.string,
-        samlLoginButtonText: PropTypes.string,
-        siteName: PropTypes.string,
-        initializing: PropTypes.bool,
-        actions: PropTypes.shape({
-            login: PropTypes.func.isRequired,
-            addUserToTeamFromInvite: PropTypes.func.isRequired,
-        }).isRequired,
+interface Props extends RouteProps {
+    intl: IntlShape;
+    isLicensed: boolean;
+    currentUser?: UserProfile;
+    customBrandText?: string;
+    customDescriptionText?: string
+    enableCustomBrand: boolean
+    enableLdap: boolean;
+    enableOpenServer: boolean;
+    enableSaml: boolean;
+    enableSignInWithEmail: boolean;
+    enableSignInWithUsername: boolean;
+    enableSignUpWithEmail: boolean;
+    enableSignUpWithGitLab: boolean;
+    enableSignUpWithGoogle: boolean;
+    enableSignUpWithOffice365: boolean;
+    experimentalPrimaryTeam?: string;
+    ldapLoginFieldName?: string;
+    samlLoginButtonText?: string;
+    siteName?: string;
+    initializing?: boolean;
+    actions: {
+        login: (loginId: string, password: string, mfaToken: string) => Promise<{ data: boolean, error: ServerError }>;
+        addUserToTeamFromInvite: (token: string, inviteId: string) => Promise<{ data: Team, error: ServerError }>
     }
+}
 
-    constructor(props) {
+type State = {
+    ldapEnabled: boolean;
+    usernameSigninEnabled: boolean;
+    emailSigninEnabled: boolean;
+    samlEnabled: boolean
+    loginId: string;
+    password: string;
+    showMfa: boolean;
+    loading: boolean;
+    sessionExpired: boolean;
+    brandImageError: boolean;
+    serverError: JSX.Element;
```

Comment:
If we use `React.ReactNode` here instead, we don't need to define a blank `<div/>` below.